[SYCL][Matrix] Add documentation about new matrix features (#6157)
intel/llvm: sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix.asciidoc

The matrix multiply and add function performs the multiply operation on the matrices `A` and `B`, accumulate the result with `C` and return the result.

#### Matrix Initialization: `joint_matrix_fill`
The current interface presented above assumes that all the matrices are directly loaded from memory. This new function called `joint_matrix_fill`  makes it possible to multiply a matrix which is not directly loaded from memory but rather initialized directly in the register. On Intel AMX, if the initialization constant is zero, this would map to `_tile_zero` intrinsic: 

```c++
namespace sycl::ext::oneapi::experimental::matrix {
  template <typename Group, typename T, size_t NumRows, size_t NumCols,
          matrix_layout L, typename Tv>
  void joint_matrix_fill(Group sg, joint_matrix<T, NumRows, NumCols, L, Group> &m, const Tv v);
}
```
IMPORTANT: In the current implementation, only the subgroup scope is supported.  

#### Element Indexing and Piece-Wise Operations 

> Besides matrix multiply and add, matrices are used in linear and non linear piece-wise operations. Activation functions are an example of element-wise operations. They can be linear like `ReLU` that, for each value `z`, returns the maximum between `z` and zero, or non linear like `Sigmoid` that calculates `1/(1+ exp(-z))`. Quantization that is needed for conversion between low precision types like `int8_t` and `fp32` uses piece-wise operations. For instance, quantized GEMM for `int8_t` is calculated using `A*B + sum_rows_A + sum_cols_B + scalar_zero_point`. `sum_rows_A` and `sum_cols_B` do not operate on elements of the matrix but on pieces: row in `sum_rows_A` and columns in `sum_cols_B`.

Author: There is that: element-wise operations where  computation needs to be performed individually on each element of the matrix. But there is also piece-wise operations like sum of rows. An example on how to do that using get_wi_data can be:

auto data = C.get_wi_data();
for (int i = 0; i < length; ++i) {
  auto row, col = data[i].get_coord();
  sum_of_local_rows[row] += data[i];
}

But in this case, the user needs to know the mapping WI data to matrix first. As you can see here, the mapping function is missing.  I added in the future looking API the mapping function as future work.

In order to be able to perform any piece-wise operation in a general and in an efficient way, we provide a mapping conversion function from the matrix domain that is owned by a group of work items to the portion that is owned by a work item. Besides, the WI data to joint matrix mapping coordinates information must be known. 

Nvidia wmma interface added a new member to `fragment` class to designate the WI owned part of the matrix. 
While this provides fast element indexing on the GPU compared to the non-restrictive option, the user does not know the mapping of the owned data to the original matrix. 
 However using the `mma` ptx instructions as opposed to the `wmma` ptx instructions the mapping is known. Knowing this mapping is important for the user to implement new operations like sum of rows of a matrix for quantized algorithms.

> ##### Solution: Explicit conversion with mapping from SIMD to SPMD

Author: I will move these in a background subsection but I will leave it in this section. Let's see if it looks better.

> We introduce a new function `get_wi_data` that provides a view of the portion of the matrix that is owned by the current WI.

Author: I already added more clarifications based on Jack's first review. I will add more based on your input as well. However, note that , "such that each work-item owns a unique subset of the data elements"  is not always true like in the AMX case for instance. A matrix is allocated in the 2d register tile that is a subgroup shared memory (register in this case).

> We introduce a new function `get_wi_data` that provides a view of the portion of the matrix that is owned by the current WI.

Author: >>> Are you saying that when one work-item calls get_wi_data that it might get overlapping elements 

No this is not possible

> ```c++

Reviewer: The purpose of these code synopses is to show the API, not the implementation.  Therefore, remove the function bodies and all the private data members.  For example:

```
namespace sycl::ext::oneapi::experimental::matrix {

template <typename T, size_t NumRows, size_t NumCols,
          matrix_layout Layout = matrix_layout::row_major,
          typename Group = sycl::sub_group>
struct joint_matrix {
   wi_data<T, NumRows, NumCols, Layout, Group> get_wi_data();
};

/* ... */

} // namespace sycl::ext::oneapi::experimental::matrix
```

Author: Will do that, thanks

> ```c++

Reviewer: The purpose of these code synopses is to show the API, not the implementation.  Therefore, remove the function bodies and all the private data members.  For example:

```
namespace sycl::ext::oneapi::experimental::matrix {

template <typename T, size_t NumRows, size_t NumCols,
          matrix_layout Layout = matrix_layout::row_major,
          typename Group = sycl::sub_group>
struct joint_matrix {
   wi_data<T, NumRows, NumCols, Layout, Group> get_wi_data();
};

/* ... */

} // namespace sycl::ext::oneapi::experimental::matrix
```

namespace sycl::ext::oneapi::experimental::matrix {
template <typename T, size_t NumRows, size_t NumCols,
          matrix_layout Layout = matrix_layout::row_major,
          typename Group = sycl::sub_group>
struct joint_matrix {
   joint_matrix(Group sg) {…}   
   wi_data<T, NumRows, NumCols, Layout, Group> get_wi_data() {
     return wi_data<T, NumRows, NumCols, Layout, Group>(*this);
   }
};

> template <typename T, size_t NumRows, size_t NumCols, matrix_layout Layout, typename Group>

Reviewer: Do `wi_data` and `wi_element` really need all these template parameters?  It seems like it would be easier to use if the only template parameter was `T`.  It seems like the other template parameters are only there because there is a private data member `M` (a reference to matrix).  However, you only seem to use `M.spvm` in the function bodies.  Could you instead just store the `spvm` member directly in `wi_data` and `wi_element`?

Author: Can you elaborate on your suggestion? I get what you want to do but did not get the how.

Reviewer: Sorry for the delayed response.  I was OOO for about 2 months and just got back recently.

I was thinking that these types could be simplified to have fewer template parameters, so the goal would be to have an API like this:

```
template <typename T>
class wi_data {
  size_t length();
  wi_element<T> operator[](size_t i);
};

template <typename T>
class wi_element {
  operator T();
  wi_element &operator=(const T &rhs);
};
```

Very roughly, I was thinking that you could accomplish this by changing the private data member included in `wi_data` and `wi_element`.  Currently, these both contain a reference to the joint matrix `M`.  However, it seems like they only need to use `M.spvm`.  Therefore, could you change the implementation to hold just the `spvm` like:

```
template <typename T>
class wi_data {
  /* not sure what type */ spvm;
public:
  size_t length() {return __spirv_JointMatrixWorkItemLengthINTEL(spvm);}
  wi_element<T> operator[](size_t i) {
    return wi_element<T>(spvm, i);
  }
};

template <typename T>
class wi_element {
  /* not sure what type */ spvm;
  std::size_t idx;

public:
  operator T() {
    return __spirv_VectorExtractDynamic(spvm, idx);
  }
  wi_element &operator=(const T &rhs) {
    M.spvm = __spirv_VectorInsertDynamic(spvm, rhs, idx);
    return *this;
  }
};
```

Author: @gmlueck 
It looks like spvm type also needs these template parameters. SO I don't think we can reduce them:
 __spv::__spirv_JointMatrixINTEL<
      T, NumRows, NumCols, spv_matrix_layout_traits<Layout>::value> *spvm;

Reviewer: That's too bad.

> template <typename T, size_t NumRows, size_t NumCols, matrix_layout Layout, typename Group>

Reviewer: Do `wi_data` and `wi_element` really need all these template parameters?  It seems like it would be easier to use if the only template parameter was `T`.  It seems like the other template parameters are only there because there is a private data member `M` (a reference to matrix).  However, you only seem to use `M.spvm` in the function bodies.  Could you instead just store the `spvm` member directly in `wi_data` and `wi_element`?

class wi_data {
  joint_matrix<T, NumRows, NumCols, Layout, Group> &M;
public:

>   wi_data(joint_matrix<T, NumRows, NumCols, Layout, Group> &Mat) : M(Mat){}

Reviewer: I presume we do NOT want application code to construct a `wi_data` (or `wi_element`) directly?  Instead, I presume we want application to call `joint_matrix::get_wi_data` to get the `wi_data`?  If that is the case, these constructors should be private in the implementation, and `joint_matrix` should be a friend, so that it can construct the objects.

The code synopsis, then, would only list the public member functions:

```
template <typename T, size_t NumRows, size_t NumCols, matrix_layout Layout, typename Group>
class wi_data {
 public:
  size_t length();
  wi_element<T, NumRows, NumCols, Layout, Group> operator[](size_t i);
};

template <typename T, size_t NumRows, size_t NumCols, matrix_layout Layout, typename Group>
class wi_element {
 public:
  operator T();
  wi_element &operator=(const T &rhs);
};
```

>   wi_data(joint_matrix<T, NumRows, NumCols, Layout, Group> &Mat) : M(Mat){}

Reviewer: I presume we do NOT want application code to construct a `wi_data` (or `wi_element`) directly?  Instead, I presume we want application to call `joint_matrix::get_wi_data` to get the `wi_data`.  If that is the case, these constructors should be private in the implementation, and `joint_matrix` should be a friend, so that it can construct the objects.

The coded synopsis, then, would only list the public member functions:

```
template <typename T, size_t NumRows, size_t NumCols, matrix_layout Layout, typename Group>
class wi_data {
 public:
  size_t length();
  wi_element<T, NumRows, NumCols, Layout, Group> operator[](size_t i);
};
```

  size_t length() {return __spirv_JointMatrixWorkItemLengthINTEL(M.spvm);}
  wi_element<T, NumRows, NumCols, Layout, Group> operator[](size_t i) {
    return wi_element<T, NumRows, NumCols, Layout, Group>(M, i);
  }
};
template <typename T, size_t NumRows, size_t NumCols,
          matrix_layout Layout = matrix_layout::row_major,
          typename Group = sycl::sub_group>
class wi_element {
  joint_matrix<T, NumRows, NumCols, Layout, Group> &M;
  std::size_t idx;

public:
  wi_element(joint_matrix<T, NumRows, NumCols, Layout, Group> &Mat, std::size_t i): M(Mat), idx(i) {}
  operator T() {
    return __spirv_VectorExtractDynamic(M.spvm, idx);
  }
  wi_element &operator=(const T &rhs) {
    M.spvm = __spirv_VectorInsertDynamic(M.spvm, rhs, idx);
    return *this;
  }
…
};
}

> ```

Reviewer: After the code synopsis, there should be some description of the member functions.  I'd suggest three tables, one for each class:

* Table describing member functions of `joint_matrix` (`get_wi_data`)
* Table describing member functions of `wi_data`
* Table describing member functions of `wi_element`.

You can see an example here: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_device_global.asciidoc#representation-of-device-globals

(Scroll down to the table after the code synopsis.)

Author: since we  don't have that many members, I added description in the text. Let me know if it looks enough.

> ```

Reviewer: After the code synopsis, there should be some description of the member functions.  I'd suggest three tables, one for each class:

* Table describing member functions of `joint_matrix` (`get_wi_data`)
* Table describing member functions of `wi_data`
* Table describing member functions of `wi_element`.

You can see an example here: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_device_global.asciidoc#representation-of-device-globals

(Scroll down to the table after the code synopsis.)

Example where each WI gets a vector in `wi_data_c`. Vectorization along the subgroup dimension will get enabled automatically to vectorize the contiguous portion of the matrix.  
```c++
auto wi_data_c = matC.get_wi_data();             
for (int i = 0; i < wi_data_c.length(); i++)                

>         wi_data_c[i] *= alpha;    // Note that the indexing here “i” is in the vector owned by a WI, not in the matrix C        

Reviewer: ```suggestion
        wi_data_c[i] *= alpha;    // Note that the indexing here "i" is in the vector owned by a WI, not in the matrix C        
```

>         wi_data_c[i] *= alpha;    // Note that the indexing here “i” is in the vector owned by a WI, not in the matrix C        

Reviewer: ```suggestion
        wi_data_c[i] *= alpha;    // Note that the indexing here "i" is in the vector owned by a WI, not in the matrix C        
```

#### Matrix Initialization: `joint_matrix_fill`
The current interface presented above assumes that all the matrices are directly loaded from memory. This new function called `joint_matrix_fill`  makes it possible to multiply a matrix which is not directly loaded from memory but rather initialized directly in the register. On Intel AMX, if the initialization constant is zero, this would map to `_tile_zero` intrinsic: 

```c++
namespace sycl::ext::oneapi::experimental::matrix {
  template <typename Group, typename T, size_t NumRows, size_t NumCols,
          matrix_layout L, typename Tv>
  void joint_matrix_fill(Group sg, joint_matrix<T, NumRows, NumCols, L, Group> &m, Tv v);
}
```
IMPORTANT: In the current implementation, only the subgroup scope is supported.  

#### Element Indexing and Piece-Wise Operations 
##### Background
Besides matrix multiply and add, this extension aims to make it possible to perform piece-wise operations on matrices in a SPMD manner. The mechanisms that are recommended to perform such piece-wise operations depend upon which of the following classes the operation falls into:

> Class 1- Element-wise operations where the same operation is performed on every element of the matrix. Activation functions are an example of element-wise operations.

Reviewer 2: ```suggestion
Class 1- Unary element-wise operations where the same operation is performed on every element of the matrix, such that the operation can be performed without knowledge of the position of the element within the matrix. Activation functions or adding a constant value to every element of the matrix are two examples.
```

The matrix multiply and add function performs the multiply operation on the matrices `A` and `B`, accumulate the result with `C` and return the result.

#### Matrix Initialization: `joint_matrix_fill`
The current interface presented above assumes that all the matrices are directly loaded from memory. This new function called `joint_matrix_fill`  makes it possible to multiply a matrix which is not directly loaded from memory but rather initialized directly in the register. On Intel AMX, if the initialization constant is zero, this would map to `_tile_zero` intrinsic: 

```c++
namespace sycl::ext::oneapi::experimental::matrix {
  template <typename Group, typename T, size_t NumRows, size_t NumCols,
          matrix_layout L, typename Tv>
  void joint_matrix_fill(Group sg, joint_matrix<T, NumRows, NumCols, L, Group> &m, const Tv v);
}
```
IMPORTANT: In the current implementation, only the subgroup scope is supported.  

#### Element Indexing and Piece-Wise Operations 

> Besides matrix multiply and add, matrices are used in linear and non linear piece-wise operations. Activation functions are an example of element-wise operations. They can be linear like `ReLU` that, for each value `z`, returns the maximum between `z` and zero, or non linear like `Sigmoid` that calculates `1/(1+ exp(-z))`. Quantization that is needed for conversion between low precision types like `int8_t` and `fp32` uses piece-wise operations. For instance, quantized GEMM for `int8_t` is calculated using `A*B + sum_rows_A + sum_cols_B + scalar_zero_point`. `sum_rows_A` and `sum_cols_B` do not operate on elements of the matrix but on pieces: row in `sum_rows_A` and columns in `sum_cols_B`.

Reviewer 2: I think all that needs to be stated is that:

```suggestion
Besides matrix multiply and add, this extension aims to make it possible to perform piece-wise operations on matrices in a SPMD manner. The mechanisms used to perform such piece-wise operations depend upon whether the operation falls into which of the following two classes:
1. Element-wise operations that are performed identically on every element of the matrix.
2. Element-wise operations that depend on the element index of the matrix or operations that take multiple elements as operands (such as a sum of all elements in a row for example).

This extension currently only supports case 1). However a proposal for supporting 2) in the future is provided in a later section.
```

Then continue with the explanation of how case 1) is dealt with. Case 2) seems to have been considered in section "### WI data to joint matrix mapping coordinates information for piece-wise operations"

Reviewer 2: Basically case 1) doesn't require mapping between get_data and joint_matrix, but cases 2) do. The requirement of the mapping is the distinguishing factor that determines the interfaces we use to perform the op.

Author: I will make the change, thanks

> We explored multiple options to enable this feature in the matrix interface: 1) Allowing non-restrictive element indexing on the matrix elements would result into slow indexing on the GPU, 2) Operator overloading can represent only element-wise operations and not the operations on pieces (row, column, diagonal, etc) of the matrix. 3) Providing specific functions for these piece-wise operations can resolve some of the functions we know of today like the ones involved in quantization but it is not general to any problem that may occur in the future. 

Reviewer 2: Probably this paragraph can be removed also I think.

Author: Okay

> In order to be able to perform any piece-wise operation in a general and efficient way, we provide a mapping conversion function from the matrix domain that is owned by a group of work items to the portion that is owned by each work item. This enables the WI to perform piece-wise operations on the matrix within the SYCL SPMD programming model. In joint matrix, it is up to the implementation to distribute the matrix among the work items or keep it shared. For instance, the matrix is a shared entity among the work items in the case of AMX where the AMX tile that holds the matrix data is a 2d register that is shared among the work items. Therefore the partitioning among the WIs is implementation defined.  Here, for the purpose of piece-wise operations, the conversion to the SPMD model happens using the matrix elements to WI mapping. Besides, the WI data to joint matrix mapping coordinates information must be known in order to extract the relevant piece for operations like sum of rows of a matrix. 

Reviewer 2: The AMX example is a good addition here. I think it can be made a bit more clear and precise:

```suggestion
There is no fixed allocation of matrix elements held by a `joint_matrix` instance to the WIs comprising the group used to instantiate it. For instance, the matrix is a shared entity among the WIs in the case of the AMX backend because the AMX tile that holds the matrix data is a 2d register that is shared among the work items. However it is necessary to allocate WIs to specific elements of the matrix. In order to be able to perform piece-wise operations in a general and efficient way, we provide a conversion function from the joint_matrix domain that is owned by a group of work items to the portion that is owned by each work item. This enables the WI to perform piece-wise operations on the matrix within the SYCL SPMD programming model.
```

Author: Will do, thanks

In order to be able to perform any piece-wise operation in a general and efficient way, we provide a mapping conversion function from the matrix domain that is owned by a group of work items to the portion that is owned by each work item. This enables the WI to perform piece-wise operations on the matrix within the SYCL SPMD programming model. In joint matrix, it is up to the implementation to distribute the matrix among the work items or keep it shared. For instance, the matrix is a shared entity among the work items in the case of AMX where the AMX tile that holds the matrix data is a 2d register that is shared among the work items. Therefore the partitioning among the WIs is implementation defined.  Here, for the purpose of piece-wise operations, the conversion to the SPMD model happens using the matrix elements to WI mapping. Besides, the WI data to joint matrix mapping coordinates information must be known in order to extract the relevant piece for operations like sum of rows of a matrix. 

> Nvidia wmma interface added a new member to `fragment` class to designate the WI owned part of the matrix. 

Reviewer 2: It would be good to remove this whole paragraph because it (or something similar) is more appropriate for the CUDA backend spec.

Author: Okay

}
```

> Example where each WI gets a vector in `wi_data_c`. Vectorization along the subgroup dimension will get enabled automatically to vectorize the contiguous portion of the matrix.  Since `wi_data_c` constructs a view of the joint matrix `matC`, `wi_data_c[i] OP rhs` updates the corresponding matrix element in the joint_matrix `matC`.

Reviewer 2: This sentence:

"Vectorization along the subgroup dimension will get enabled automatically to vectorize the contiguous portion of the matrix."

implies to me that WI portion of joint_matrix represents contiguous elements of the matrix (either row or column major), but this is not generally true: In the Tensor Cores backend it is rarely true: in most cases the WI owned portion of joint_matrix includes elements that are not contiguous. I'm not sure if there is any value to this sentence and I think it can be removed. Instead there could be:

```suggestion
In the following example `wi_data_c` is a reference to the WI owned portion of the joint matrix `matC`. As such `wi_data_c[i] OP rhs` updates the corresponding matrix element in the joint_matrix `matC`.
```

Author: It does not imply WI portion of joint matrix are contiguous. It implies round robin distribution among the WIs.
Let's take a matrix example of 3 rows x 2 cols, if we assume we have SG size of 2 just for illustration here. To ensure vectorization along the SG dimension, each WI should own 1 column. so you have simd of 2 (1 row) in this case and 3 vector instructions (for each row). Does this make sense?
Does the CUDA backend generate vector code for the loop around WI portions?

IMPORTANT: The WI data to joint matrix mapping coordinates information is not implemented yet. 

> IMPORTANT: Since the current tensorcores implementation is AOT, it is possible to know how many elements are owned by each WI at compile time. so `wi_data` in this case can be of type `marray`. An additional interface will be provided for the tensorcores AOT backend. 

Reviewer 2: Clarification: 

The current Tensor Cores implementation isn't AOT: it is JIT just at a lower level than DPC++: DPC++ compiles down to CUDA "PTX" asm, which can then be compiled down to a lower level asm called "SASS" at execution time. Despite this, the number of data elements in joint_matrix owned by each WI is known at compile time for all compilation versions. The thing that isn't known at compile time is the mapping between these WI data elements and the subgroup matrix owned by joint_matrix as a whole.
The other point is that from the available information in this doc I don't see that it follows that the WI owned elements can only be returned as an array _if and only if_ "the number of data elements in joint_matrix owned by each WI is known at compile time". But this point is not important to dwell on because I think such an explanation is not really necessary for this "IMPORTANT" note: the point here is to make the user aware that a limited number of backends (Tensor Cores only atm) can return the WI portion of the matrix as an marray (although as specified earlier in the document the order of the marray elements does not have a defined mapping to the order of the matrix owned by the joint_matrix subgroup as a whole).

```suggestion
IMPORTANT: The Tensor Cores implementation is capable of returning the complete set of matrix elements owned by a WI as an `marray`, which can be useful in cases where the user wishes to perform identical operations on every element of the matrix efficiently and conveniently, where SYCL math functions optimized for `marray` exist. An additional interface will be provided for the Tensor Cores backend to return the `marray`.
```

Author: When I say AOT, I am referring to the SYCL compilation flow . When users have to specify -fsycl-targets this is AOT. JIT is when the compiler generates SPIRV. At runtime SPIRV is translated to the target-specific offload binaries.

I will add a comment about the fact that get_wi_data length is  a runtime variable because:
1- SYCL JIT compilation and partitioning among WIs is implementation defined 
2- SG size is not fixed (like in the CUDA backend where warp size is always 32)

In the case of CUDA backend which is SYCL AOT compiled and SG size = 32 known and fixed, the additional marray capability will be provided.
